File: Backend/app/schemas/ticket.py
# Ticket tiers, price preview, purchase, and discounts.
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, field_validator, model_validator


class TicketTierCreate(BaseModel):
    name: str
    description: str | None = None
    price_cents: int
    max_reserved_spots: int = 0
    display_order: int = 0

    @field_validator("price_cents")
    @classmethod
    def price_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("price_cents must be >= 0")
        return v

    @field_validator("max_reserved_spots")
    @classmethod
    def reserved_non_negative(cls, v: int) -> int:
        if v < 0:
            raise ValueError("max_reserved_spots must be >= 0")
        return v


class TicketTierUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    price_cents: int | None = None
    max_reserved_spots: int | None = None
    display_order: int | None = None

    @field_validator("price_cents")
    @classmethod
    def price_non_negative(cls, v: int | None) -> int | None:
        if v is not None and v < 0:
            raise ValueError("price_cents must be >= 0")
        return v

    @field_validator("max_reserved_spots")
    @classmethod
    def reserved_non_negative(cls, v: int | None) -> int | None:
        if v is not None and v < 0:
            raise ValueError("max_reserved_spots must be >= 0")
        return v


class TicketTierResponse(BaseModel):
    id: int
    event_id: int
    name: str
    description: str | None = None
    price_cents: int
    max_reserved_spots: int = 0
    display_order: int

    model_config = {"from_attributes": True}


class TicketPricePreviewResponse(BaseModel):
    tier_price_cents: int
    common_discount_cents: int
    selective_discount_cents: int
    pledge_discount_cents: int
    event_discount_cents: int = 0
    total_discount_cents: int
    final_price_cents: int
    commission_cents: int = 0
    net_to_organizer_cents: int = 0
    ticket_commission_percent: int = 0


class TicketPurchaseBody(BaseModel):
    tier_id: int
    quantity: int = 1  # number of tickets to purchase at once
    extra_perks: str | None = None  # when discount >= price, organizer can set this later or at purchase

    @field_validator("quantity")
    @classmethod
    def quantity_positive(cls, v: int) -> int:
        if v < 1 or v > 10:
            raise ValueError("quantity must be between 1 and 10")
        return v


class TicketSaleResponse(BaseModel):
    id: int
    event_id: int
    user_id: int
    ticket_tier_id: int
    purchase_group_id: str | None = None  # groups multi-ticket purchases
    ticket_code: str  # unique code for QR; customer displays as QR for organizer to scan
    receipt_number: str | None = None  # human-readable receipt ID
    tier_name: str | None = None
    event_title: str | None = None
    attendee_display_name: str | None = None  # name on ticket (holder's display_name or email)
    amount_paid_cents: int
    discount_applied_cents: int
    commission_cents: int = 0
    net_to_organizer_cents: int = 0
    extra_perks: str | None
    status: str
    scanned_at: datetime | None = None  # past scan time when already scanned
    scanned_by_id: int | None = None
    scanned_by_display_name: str | None = None  # who scanned it
    encrypted_qr_payload: str | None = None  # AES-256-GCM encrypted QR data
    created_at: datetime


class TicketReceiptResponse(BaseModel):
    """Full receipt detail for a ticket purchase."""
    sale_id: int
    user_id: int
    receipt_number: str
    ticket_code: str
    status: str
    # Attendee
    attendee_name: str | None = None
    # Event
    event_id: int
    event_title: str
    event_start_time: datetime | None = None
    event_end_time: datetime | None = None
    # Organizer
    organizer_name: str | None = None
    organizer_email: str | None = None
    organizer_phone: str | None = None
    # Venue
    venue_name: str | None = None
    venue_address: str | None = None
    # Tier
    tier_name: str
    tier_price_cents: int
    # Payment
    amount_paid_cents: int
    discount_applied_cents: int
    commission_cents: int = 0
    net_to_organizer_cents: int = 0
    extra_perks: str | None = None
    encrypted_qr_payload: str | None = None  # AES-256-GCM encrypted QR data
    # Timestamps
    purchased_at: datetime
    scanned_at: datetime | None = None


class ScanTicketBody(BaseModel):
    ticket_code: str | None = None          # legacy plaintext ticket code
    encrypted_payload: str | None = None    # AES-256-GCM encrypted QR blob

    @model_validator(mode="after")
    def require_at_least_one(self) -> "ScanTicketBody":
        if not self.ticket_code and not self.encrypted_payload:
            raise ValueError("Provide either ticket_code or encrypted_payload")
        return self


class ScanTicketResponse(BaseModel):
    """Scan result: already_scanned, past scan time and name on ticket are on ticket."""
    already_scanned: bool
    ticket: TicketSaleResponse  # includes scanned_at (past scan time), attendee_display_name (name on ticket)


class TicketSummaryItem(BaseModel):
    """Individual ticket within a purchase group."""
    sale_id: int
    ticket_code: str
    receipt_number: str | None = None
    encrypted_qr_payload: str | None = None
    status: str
    scanned_at: datetime | None = None


class PurchaseGroupReceiptResponse(BaseModel):
    """Aggregated receipt for a multi-ticket purchase group."""
    purchase_group_id: str
    # Event
    event_id: int
    event_title: str
    event_start_time: datetime | None = None
    event_end_time: datetime | None = None
    # Organizer
    organizer_name: str | None = None
    organizer_email: str | None = None
    organizer_phone: str | None = None
    # Venue
    venue_name: str | None = None
    venue_address: str | None = None
    # Attendee
    attendee_name: str | None = None
    # Tier
    tier_name: str
    tier_price_cents: int
    # Aggregated totals
    quantity: int
    total_amount_paid_cents: int
    total_discount_applied_cents: int
    total_commission_cents: int = 0
    total_net_to_organizer_cents: int = 0
    # Tickets
    tickets: list[TicketSummaryItem]
    # Timestamps
    purchased_at: datetime


class TicketSalesStatsResponse(BaseModel):
    """Stats for ticket sales (scanned/total) on an event."""
    total_sold: int
    total_scanned: int


class UserDiscountBody(BaseModel):
    user_id: int
    discount_type: Literal["percent", "fixed_cents"]
    value: int  # 0-100 for percent, or cents for fixed_cents

File: FrontEnd/lib/models/ticket.dart
// ─── Ticket Strategy Request Models ───

class CreateTicketStrategyTierInput {
  final String name;
  final String? description;
  final int priceCents;
  final int displayOrder;

  const CreateTicketStrategyTierInput({
    required this.name,
    this.description,
    required this.priceCents,
    this.displayOrder = 0,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'price_cents': priceCents,
        'display_order': displayOrder,
        if (description != null && description!.isNotEmpty) 'description': description,
      };
}

class CreateTicketStrategyRequest {
  final String name;
  final List<CreateTicketStrategyTierInput> tiers;

  const CreateTicketStrategyRequest({
    required this.name,
    required this.tiers,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'tiers': tiers.map((t) => t.toJson()).toList(),
      };
}

class UpdateTicketStrategyRequest {
  final String? name;
  final List<CreateTicketStrategyTierInput>? tiers;

  const UpdateTicketStrategyRequest({this.name, this.tiers});

  Map<String, dynamic> toJson() => {
        if (name != null) 'name': name,
        if (tiers != null) 'tiers': tiers!.map((t) => t.toJson()).toList(),
      };
}

// ─── Ticket Tier Request Models ───

class CreateTicketTierRequest {
  final String name;
  final int priceCents;
  final int displayOrder;
  final String? description;
  final int? maxReservedSpots;
  final bool isFeatured;

  const CreateTicketTierRequest({
    required this.name,
    required this.priceCents,
    this.displayOrder = 0,
    this.description,
    this.maxReservedSpots,
    this.isFeatured = false,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'price_cents': priceCents,
        'display_order': displayOrder,
        if (description != null && description!.isNotEmpty) 'description': description,
        if (maxReservedSpots != null && maxReservedSpots! > 0) 'max_reserved_spots': maxReservedSpots,
        'is_featured': isFeatured,
      };
}

class UpdateTicketTierRequest {
  final String? name;
  final int? priceCents;
  final String? description;
  final int? maxReservedSpots;
  final bool? isFeatured;

  const UpdateTicketTierRequest({
    this.name,
    this.priceCents,
    this.description,
    this.maxReservedSpots,
    this.isFeatured,
  });

  Map<String, dynamic> toJson() => {
        if (name != null) 'name': name,
        if (priceCents != null) 'price_cents': priceCents,
        if (description != null) 'description': description,
        if (maxReservedSpots != null) 'max_reserved_spots': maxReservedSpots,
        if (isFeatured != null) 'is_featured': isFeatured,
      };
}

class TicketTier {
  final int id;
  final int eventId;
  final String name;
  final String? description;
  final int priceCents;
  final int maxReservedSpots;
  final int ticketsSold;
  final int spotsReserved;
  final int displayOrder;
  final bool fromStrategy;
  final bool isFeatured;

  TicketTier({
    required this.id,
    required this.eventId,
    required this.name,
    this.description,
    required this.priceCents,
    this.maxReservedSpots = 0,
    this.ticketsSold = 0,
    this.spotsReserved = 0,
    required this.displayOrder,
    this.fromStrategy = false,
    this.isFeatured = false,
  });

  factory TicketTier.fromJson(Map<String, dynamic> json) {
    return TicketTier(
      id: (json['id'] as num?)?.toInt() ?? 0,
      eventId: (json['event_id'] as num?)?.toInt() ?? 0,
      name: json['name'] as String? ?? '',
      description: json['description'] as String?,
      priceCents: (json['price_cents'] as num?)?.toInt() ?? 0,
      maxReservedSpots: json['max_reserved_spots'] ?? 0,
      ticketsSold: json['tickets_sold'] ?? 0,
      spotsReserved: json['spots_reserved'] ?? 0,
      displayOrder: json['display_order'] ?? 0,
      fromStrategy: json['from_strategy'] ?? false,
      isFeatured: json['is_featured'] ?? false,
    );
  }

  bool get isFree => priceCents == 0;
  String get priceFormatted => isFree ? 'FREE' : '\$${(priceCents / 100).toStringAsFixed(2)}';

  int get spotsLeft => maxReservedSpots > 0
      ? (maxReservedSpots - ticketsSold - spotsReserved)
      : 0;
}

class TicketSale {
  final int id;
  final int eventId;
  final int userId;
  final int ticketTierId;
  final String? purchaseGroupId;
  final String ticketCode;
  final String? receiptNumber;
  final String? tierName;
  final String? eventTitle;
  final String? eventStatus;
  final String? attendeeDisplayName;
  final String? scannedByDisplayName;
  final int amountPaidCents;
  final int discountAppliedCents;
  final int commissionCents;
  final int netToOrganizerCents;
  final String? extraPerks;
  final String status;
  final DateTime? scannedAt;
  final String? encryptedQrPayload;
  final DateTime createdAt;

  TicketSale({
    required this.id,
    required this.eventId,
    required this.userId,
    required this.ticketTierId,
    this.purchaseGroupId,
    required this.ticketCode,
    this.receiptNumber,
    this.tierName,
    this.eventTitle,
    this.eventStatus,
    this.attendeeDisplayName,
    this.scannedByDisplayName,
    required this.amountPaidCents,
    required this.discountAppliedCents,
    this.commissionCents = 0,
    this.netToOrganizerCents = 0,
    this.extraPerks,
    required this.status,
    this.scannedAt,
    this.encryptedQrPayload,
    required this.createdAt,
  });

  factory TicketSale.fromJson(Map<String, dynamic> json) {
    return TicketSale(
      id: (json['id'] as num?)?.toInt() ?? 0,
      eventId: (json['event_id'] as num?)?.toInt() ?? 0,
      userId: (json['user_id'] as num?)?.toInt() ?? 0,
      ticketTierId: (json['ticket_tier_id'] as num?)?.toInt() ?? 0,
      purchaseGroupId: json['purchase_group_id'] as String?,
      ticketCode: json['ticket_code'] as String? ?? '',
      receiptNumber: json['receipt_number'] as String?,
      tierName: json['tier_name'] as String?,
      eventTitle: json['event_title'] as String?,
      eventStatus: json['event_status'] as String?,
      attendeeDisplayName: json['attendee_display_name'] as String?,
      scannedByDisplayName: json['scanned_by_display_name'] as String?,
      amountPaidCents: (json['amount_paid_cents'] as num?)?.toInt() ?? 0,
      discountAppliedCents: json['discount_applied_cents'] ?? 0,
      commissionCents: json['commission_cents'] ?? 0,
      netToOrganizerCents: json['net_to_organizer_cents'] ?? 0,
      extraPerks: json['extra_perks'] as String?,
      status: json['status'] as String? ?? '',
      scannedAt: json['scanned_at'] != null
          ? DateTime.parse(json['scanned_at'])
          : null,
      encryptedQrPayload: json['encrypted_qr_payload'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'event_id': eventId,
        'user_id': userId,
        'ticket_tier_id': ticketTierId,
        if (purchaseGroupId != null) 'purchase_group_id': purchaseGroupId,
        'ticket_code': ticketCode,
        if (receiptNumber != null) 'receipt_number': receiptNumber,
        if (tierName != null) 'tier_name': tierName,
        if (eventTitle != null) 'event_title': eventTitle,
        if (eventStatus != null) 'event_status': eventStatus,
        if (attendeeDisplayName != null) 'attendee_display_name': attendeeDisplayName,
        if (scannedByDisplayName != null) 'scanned_by_display_name': scannedByDisplayName,
        'amount_paid_cents': amountPaidCents,
        'discount_applied_cents': discountAppliedCents,
        'commission_cents': commissionCents,
        'net_to_organizer_cents': netToOrganizerCents,
        if (extraPerks != null) 'extra_perks': extraPerks,
        'status': status,
        if (scannedAt != null) 'scanned_at': scannedAt!.toIso8601String(),
        if (encryptedQrPayload != null) 'encrypted_qr_payload': encryptedQrPayload,
        'created_at': createdAt.toIso8601String(),
      };

  String get amountPaidFormatted =>
      '\$${(amountPaidCents / 100).toStringAsFixed(2)}';

  bool get isScanned => scannedAt != null;
}

class TicketPricePreview {
  final int tierPriceCents;
  final int commonDiscountCents;
  final int selectiveDiscountCents;
  final int pledgeDiscountCents;
  final int eventDiscountCents;
  final int milestoneDiscountCents;
  final int earlyBirdDiscountCents;
  final int totalDiscountCents;
  final int finalPriceCents;
  final int commissionCents;
  final int netToOrganizerCents;
  final int ticketCommissionPercent;
  final bool discountCapped;
  final int maxDiscountPercent;

  TicketPricePreview({
    required this.tierPriceCents,
    this.commonDiscountCents = 0,
    this.selectiveDiscountCents = 0,
    this.pledgeDiscountCents = 0,
    this.eventDiscountCents = 0,
    this.milestoneDiscountCents = 0,
    this.earlyBirdDiscountCents = 0,
    this.totalDiscountCents = 0,
    required this.finalPriceCents,
    this.commissionCents = 0,
    this.netToOrganizerCents = 0,
    this.ticketCommissionPercent = 0,
    this.discountCapped = false,
    this.maxDiscountPercent = 100,
  });

  factory TicketPricePreview.fromJson(Map<String, dynamic> json) {
    return TicketPricePreview(
      tierPriceCents: json['tier_price_cents'] as int? ?? 0,
      commonDiscountCents: json['common_discount_cents'] as int? ?? 0,
      selectiveDiscountCents: json['selective_discount_cents'] as int? ?? 0,
      pledgeDiscountCents: json['pledge_discount_cents'] as int? ?? 0,
      eventDiscountCents: json['event_discount_cents'] as int? ?? 0,
      milestoneDiscountCents: json['milestone_discount_cents'] as int? ?? 0,
      earlyBirdDiscountCents: json['early_bird_discount_cents'] as int? ?? 0,
      totalDiscountCents: json['total_discount_cents'] as int? ?? 0,
      finalPriceCents: json['final_price_cents'] as int? ?? 0,
      commissionCents: json['commission_cents'] as int? ?? 0,
      netToOrganizerCents: json['net_to_organizer_cents'] as int? ?? 0,
      ticketCommissionPercent: json['ticket_commission_percent'] as int? ?? 0,
      discountCapped: json['discount_capped'] as bool? ?? false,
      maxDiscountPercent: json['max_discount_percent'] as int? ?? 100,
    );
  }

  String get finalPriceFormatted =>
      '\$${(finalPriceCents / 100).toStringAsFixed(2)}';
}

class TicketSalesStats {
  final int totalSold;
  final int totalScanned;

  TicketSalesStats({
    required this.totalSold,
    required this.totalScanned,
  });

  factory TicketSalesStats.fromJson(Map<String, dynamic> json) {
    return TicketSalesStats(
      totalSold: json['total_sold'] as int? ?? 0,
      totalScanned: json['total_scanned'] as int? ?? 0,
    );
  }
}

class TicketScanResult {
  final bool alreadyScanned;
  final TicketSale ticket;

  TicketScanResult({
    required this.alreadyScanned,
    required this.ticket,
  });

  factory TicketScanResult.fromJson(Map<String, dynamic> json) {
    return TicketScanResult(
      alreadyScanned: json['already_scanned'] as bool? ?? false,
      ticket: TicketSale.fromJson(json['ticket'] as Map<String, dynamic>),
    );
  }
}
